src/assets/Debugging/debugging.md

```diff
@@ -0,0 +1,378 @@
+
+```bash
+docker compose exec nginx ping nodejs-app
+```
+
+**Example Output:**
+
+```console
+PING nodejs-app (172.19.0.3): 56 data bytes
+64 bytes from 172.19.0.3: seq=0 ttl=64 time=0.042 ms
+```
+
+### Examining Post Accessibility
+Check if Node.js is listening on port 3000 inside its container:
+
+```bash
+docker exec nodejs-app netstat -tuln
+```
+
+- `docker exec`: Runs a command inside a specific container (`nodejs-app`).
+- `netstat -tuln`: A Linux utility to list all listening network ports with the flags:
+  - `-t`: Show TCP ports
+  - `-u`: Show UDP ports
+  - `-l`: Display only listening ports (services accepting connections)
+  - `-n`: Show numeric addresses/ports
+
+**Expected Output:**
+```console
+Active Internet connections (only servers)
+Proto Recv-Q Send-Q Local Address           Foreign Address         State
+tcp        0      0 :::3000                 :::*                    LISTEN
+```
+
+- Shows Node.js listening on port 3000
+- No output means service isn't running properly
+
+## 4. Interactive Debugging
+### Executing into Containers
+
+For Node.js application debugging:
+```bash
+docker exec -it nodejs-app sh
+```
+
+This gives you full shell access inside the container where you can run any Linux command (as long as the tool exists in the container). Following are few things that you can do.
+
+1. Check running processes: `ps aux`
+```bash
+ps aux
+```
+
+**Output:**
+
+```console
+PID   USER     TIME  COMMAND
+    1 root      0:00 {MainThread} node app.js
+   28 root      0:00 sh
+   45 root      0:00 ps aux
+```
+- Shows all running processes in the container
+- `PID1`: Your Node.js application (node app.js)
+- `PID28`: The shell session you just started
+- `PID 45`: The `ps aux` command itself
+- Confirms your application is running as the main process
+
+2. Examine environment variables: `printenv`
+```bash
+printenv
+```
+
+**Output:**
+```console
+NODE_VERSION=23.11.0
+HOSTNAME=ada88201c429
+YARN_VERSION=1.22.22
+SHLVL=1
+HOME=/root
+TERM=xterm
+PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
+PWD=/app
+```
+- Displays all environment variables to your Node.js app
+
+3. Test Redis connectivity: `redis-cli -h redis ping`
+```bash
+redis-cli -h redis ping
+```
+
+**Output:**
+```console
+PONG
+```
+
+- Tests connectivity to your Redis container
+- `PONG` response confirms network connectivity
+
+### Debugging Nginx Configuration
+```bash
+docker exec -it debug-docker_nginx_1 nginx -t
+```
+
+**Expected Output:**
+```console
+nginx: the configuration file /etc/nginx/nginx.conf syntax is ok
+nginx: configuration file /etc/nginx/nginx.conf test is successful
+```
+
+## 5. Health Checks and Readiness Probes
+Docker health checks are automated tests that periodically vefity if a container is functioning properly. Health checks transform your containers from static processes into self aware services that can catch issues like application crashes, frozen processes, dependency failures, etc.
```

Comment:
```suggestion
Docker health checks are automated tests that periodically verify if a container is functioning properly. Health checks transform your containers from static processes into self aware services that can catch issues like application crashes, frozen processes, dependency failures, etc.
```

Nice, I've just learned something new. I always thought the health checks are only run on startup, but of course it makes a lot of sense that they run periodically. Thank you!